### PR TITLE
[InstCombine] Fold `X Pred C2 ? X BOp C1 : C2 BOp C1` to `min/max(X, C2) BOp C1`

### LINKED ISSUE #82414: [InstCombine] Missed optimization : fold `X > C2 ? X + C1 : C2 + C1` to `max(X, C2) + C1`
Alive2 proof: https://alive2.llvm.org/ce/z/ERjNs4

### Motivating example 

```llvm
define i32 @src(i32 %x) {
  %add = add nsw i32 %x, 16
  %cmp = icmp sgt i32 %x, 1008
  %s = select i1 %cmp, i32 %add, i32 1024
  ret i32 %s
}
```

can be folded to:

```llvm
define i32 @tgt(i32 %x) {
  %smax = call i32 @llvm.smax.i32(i32 %x, i32 1008)
  %add = add nuw nsw i32 %smax, 16
  ret i32 %add
}
```

LLVM does well when `C1` or `C2` is not constant, but when both are constants, LLVM missed it. Though this example doesn't show better codegen, I think it's a better canonicalization.

### Real-world motivation

This snippet of IR is derived from [protobuf/generated_message_tctable_lite.cc](https://github.com/dtcxzyw/llvm-opt-benchmark/blob/main/bench/protobuf/original/generated_message_tctable_lite.cc.ll@_ZN6google8protobuf13RepeatedFieldIiE14GrowNoAnnotateEii) after O3 pipeline (original IR is from [llvm-opt-benchmark](https://github.com/dtcxzyw/llvm-opt-benchmark)).
Original IR is too big to attach here, email me to get it please.

**Let me know if you can confirm that it's an optimization opportunity, thanks.**

### PR DESCRIPTION
**Pull request by veera-sivarajan**
Fixes #82414

General Proof: https://alive2.llvm.org/ce/z/ERjNs4 
Proof for Tests: https://alive2.llvm.org/ce/z/K-934G

This PR transforms `select` instructions of the form `select (Cmp X C1) (BOp X C2) C3` to `BOp (min/max X C1) C2` iff `C3 == BOp C1 C2`.

This helps in eliminating a noop loop in https://github.com/rust-lang/rust/issues/123845 but does not improve optimizations.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
Comment (veera-sivarajan) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:1932]: Not sure if flags should be copied here. alive2 shows that transformations are correct with and without flags being copied.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
Comment (veera-sivarajan) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:1933]: A binary operator is manually created and inserted because the equivalent IRBuilder method, `CreateBinOp()`, constant folds the operands.
But the operands should not be folded to allow `foldSelectOpOp()` to detect the common operand.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
Comment (veera-sivarajan) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:1936]: Two transformations, `replaceOperand()` and `foldSelectOpOp()`, are invoked in the same InstCombine loop to prevent the new binary operation with
const operands from being folded into a constant. Otherwise, `foldSelectOpOp()` would not detect the common operand.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
Comment (veera-sivarajan) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:1920]: This is quite convoluted because there's no `m_c_Select()` yet.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:1916]:
```suggestion
  if (!match(CondVal, m_ICmp(Predicate, m_Value(X), m_Constant(C1))))
```
We always canonicalize constants to the RHS.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:1923]: Please add a test for `BOp = sub C, X`.

File: llvm/test/Transforms/InstCombine/saturating-add-sub.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/saturating-add-sub.ll:1796]: Missing fold: https://alive2.llvm.org/ce/z/LnDhkk
```
define i32 @src(i32 %x, i32 %y) {
  %a = add i32 %x, -2
  %c = icmp ult i32 %x, 2
  %r = select i1 %c, i32 -1, i32 %a
  ret i32 %r
}

define i32 @tgt(i32 %x) {
  %a = call i32 @llvm.umax(i32 %x, i32 1)
  %r = add i32 %a, -2
  ret i32 %r
}
```

**Review by nikic — CHANGES_REQUESTED** [review]
Please directly create the final pattern with the min/max intrinsic. Creating temporary, non-canonical instructions and relying on another fold to deal with it is definitely going to bite us later.

File: llvm/test/Transforms/InstCombine/saturating-add-sub.ll
Comment (veera-sivarajan) [reply llvm/test/Transforms/InstCombine/saturating-add-sub.ll:1796]: Thank you for the feedback.

While this fold is similar to the one in this PR, I don't think this PR can fold the provided example because: 

In the example, `C1 = 2`, `C2 = -2`, `C3 = 1` and `BOp = add`.

This PR can fold only when `C3 == BOp C1 C2` and this does not hold true for the example.

**Comment by veera-sivarajan** [comment]
> The more idiomatic way to handle this is to have something like if (match(FalseVal, m_Constant())) { std::swap(TrueVal, FalseVal); Pred = ICmpInst::getInversePredicate(Pred); } and then you no longer have to deal with which operand the constant is in.

Thank you for the detailed feedback.

This simplifies the pattern match but I still have to conditionally assign `NewTrue` and `NewFalse` because `matchDecomposedSelectPattern()` takes an `ICmpInst` and not a `ICmpInst:Predicate`.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
Comment (veera-sivarajan) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:1919]: Thank you for the detailed feedback.

**Comment by veera-sivarajan** [comment]
> You can convert the icmp into icmp ule i32 %x, 1 with getFlippedStrictnessPredicateAndConstant.

I'm sorry, could you please explain why this is necessary there?

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:1914]: Is it always profitable to perform this transform for div/rem?

Comment (veera-sivarajan) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: No, it does not help Rust's inclusive ranges and I cannot think of any other cases where it would be always profitable.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!
Please wait for additional approval from other reviewers :)

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:1911]: This check deserves a comment -- am I understanding correctly that this is profitability heuristic, not a correctness check?

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:1919]:
```suggestion
  const DataLayout &DL = BOp->getDataLayout();
```
This one always gets the same name :)

**Review by nikic — APPROVED** [review]
LGTM, thanks!

**Comment by veera-sivarajan** [comment]
Rebased and updated the tests.

Please merge it for me. I don't have write access yet.

**Pull request merged**
